internal/generator/fri/template/fri.go.tmpl

```diff
@@ -0,0 +1,700 @@
+		tree.Push(q[i].Marshal())
+	}
+	var res OpeningProof
+	res.merkleRoot, res.proofSet, res.index, res.numLeaves = tree.Prove()
+
+	// set the claimed value, which is the first entry of the Merkle proof
+	res.ClaimedValue.SetBytes(res.proofSet[0])
+
+	return res, nil
+}
+
+func checkRoots(a, b []byte) error {
+
+	if len(a) != len(b) {
+		return ErrMerkleRoot
+	}
+	for i := 0; i < len(a); i++ {
+		if a[i] != b[i] {
+			return ErrMerkleRoot
+		}
+	}
+
+	return nil
+}
+
+// Verifies the opening of a polynomial.
+// * position the point at which the proof is opened (the point is gⁱ where i = position)
+// * openingProof Merkle path proof
+// * pp proof of proximity, needed because before opening Merkle path proof one should be sure that the
+// committed values come from a polynomial. During the verification of the Merkle path proof, the root
+// hash of the Merkle path is compared to the root hash of the first interaction of the proof of proximity,
+// those should be equal, if not an error is raised.
+func (s radixTwoFri) VerifyOpening(position uint64, openingProof OpeningProof, pp ProofOfProximity) error {
+
+	// To query the Merkle path, we look at the first series of interactions, and check whether it's the point
+	// at 'position' or its neighbor that contains the full Merkle path.
+	var fullMerkleProof int
+	if len(pp.rounds[0].interactions[0][0].proofSet) > len(pp.rounds[0].interactions[0][1].proofSet) {
+		fullMerkleProof = 0
+	} else {
+		fullMerkleProof = 1
+	}
+
+	// check that the merkle roots coincide
+	err := checkRoots(openingProof.merkleRoot, pp.rounds[0].interactions[0][fullMerkleProof].merkleRoot)
+	if err != nil {
+		return err
+	}
+
+	// convert position to the sorted version
+	sizePoly := s.domain.Cardinality
+	pos := convertCanonicalSorted(int(position), int(sizePoly))
+
+	// check the Merkle proof
+	res := merkletree.VerifyProof(s.h, openingProof.merkleRoot, openingProof.proofSet, uint64(pos), openingProof.numLeaves)
+	if !res {
+		return ErrMerklePath
+	}
+	return nil
+
+}
+
+// foldPolynomialLagrangeBasis folds a polynomial p, expressed in Lagrange basis.
+//
+// Fᵣ[X]/(Xⁿ-1) is a free module of rank 2 on Fᵣ[Y]/(Y^{n/2}-1). If
+// p∈ Fᵣ[X]/(Xⁿ-1), expressed in Lagrange basis, the function finds the coordinates
+// p₁, p₂ of p in Fᵣ[Y]/(Y^{n/2}-1), expressed in Lagrange basis. Finally, it computes
+// p₁ + x*p₂ and returns it.
+//
+// * p is the polynomial to fold, in Lagrange basis, sorted like this: p = [p(1),p(-1),p(g),p(-g),p(g²),p(-g²),...]
+// * g is a generator of the subgroup of Fᵣ^{*} of size len(p)
+// * x is the folding challenge x, used to return p₁+x*p₂
+func foldPolynomialLagrangeBasis(pSorted []fr.Element, gInv, x fr.Element) []fr.Element {
+
+	// we have the following system
+	// p₁(g²ⁱ)+gⁱp₂(g²ⁱ) = p(gⁱ)
+	// p₁(g²ⁱ)-gⁱp₂(g²ⁱ) = p(-gⁱ)
+	// we solve the system for p₁(g²ⁱ),p₂(g²ⁱ)
+	s := len(pSorted)
+	res := make([]fr.Element, s/2)
+
+	var p1, p2, twoInv, acc fr.Element
+	twoInv.SetUint64(2).Inverse(&twoInv)
```

Comment:
twoInv appears in multiple functions, maybe be worth it to initialize the constant as a global package-level variable ? (in a init() func)